ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java

```diff
@@ -0,0 +1,315 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.account.AccountUtils.*;
+import static com.github.ambry.utils.Utils.*;
+
+
+/**
+ * An implementation of {@link AccountService} that employs MySql database as its underlying storage.
+ */
+public class MySqlAccountService implements AccountService {
+
+  private MySqlAccountStore mySqlAccountStore = null;
+  private final AccountServiceMetrics accountServiceMetrics;
+  private final MySqlAccountServiceConfig config;
+  // in-memory cache for storing account and container metadata
+  private final AccountInfoMap accountInfoMap;
+  private static final Logger logger = LoggerFactory.getLogger(MySqlAccountService.class);
+  private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+  private final ScheduledExecutorService scheduler;
+
+  public MySqlAccountService(AccountServiceMetrics accountServiceMetrics, MySqlAccountServiceConfig config,
+      ScheduledExecutorService scheduler) {
+    this.accountServiceMetrics = accountServiceMetrics;
+    this.config = config;
+    this.scheduler = scheduler;
+    accountInfoMap = new AccountInfoMap(accountServiceMetrics);
+    try {
+      createMySqlAccountStore();
+    } catch (SQLException e) {
+      logger.error("MySQL account store creation failed", e);
+      // Continue account service creation. Cache will initialized with metadata from backup copy on local disk to serve read requests.
+      // Write requests will be blocked until MySql DB is up. Connection to MySql DB will be retried in polling thread that fetches new accounts.
+    }
+
+    // TODO: create backup manager to manage local back up copies of Account and container metadata and lastModifiedTime
+
+    initialFetchAndSchedule();
+
+    // TODO: Subscribe to notifications from ZK
+  }
+
+  /**
+   * creates MySql Account store which establishes connection to database
+   * @throws SQLException
+   */
+  private void createMySqlAccountStore() throws SQLException {
+    if (mySqlAccountStore == null) {
+      try {
+        mySqlAccountStore = new MySqlAccountStore(config);
+      } catch (SQLException e) {
+        // TODO: record failure, parse exception to figure out what we did wrong. If it is a non-transient error like credential issue, we should fail start up
+        logger.error("MySQL account store creation failed", e);
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Initialize in-memory cache by fetching all the {@link Account}s and {@link Container}s metadata records.
+   * It consists of 2 steps:
+   * 1. Check local disk for back up copy and load metadata and last modified time of Accounts/Containers into cache.
+   * 2. Fetch added/modified accounts and containers from mysql database since the last modified time (found in step 1)
+   *    and load into cache.
+   */
+  private void initialFetchAndSchedule() {
+
+    // TODO: Check local disk for back up copy and load metadata and last modified time into cache.
```

Comment:
Just a reminder that, for hadoop frontend, we are using rain which doesn't allow us to back up copy on disk. This can be solved by disabling backup based on`BACKUP_DIRECTORY_KEY`.  So, when backup is unavailable and MySQL connection failed, I think we should block the startup.

Reply:
@jsjtzyy Does hadoop-frontend fail to start up if it can't retrieve the account metadata from ZK?
It sounds like it will have to query MySql for the entire data set on every restart.

Reply:
I thought that the issue was just that the directory paths that the container had access to were different in that environment but reading/writing to disk was still possible?

Reply:
Oh my bad, I recall it now. We are actually using a different directory path. Sorry for misleading comment, please ignore it.